src/os/os_freertos.c

```diff
@@ -0,0 +1,117 @@
+#include "../types.h"
+#include "../oscore.h"
+#include "../main.h"
+#include "../timers.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "freertos/semphr.h"
+#include <sys/time.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdlib.h>
+
+#define STACK_DEPTH 4096
+#define TASK_PRIORITY 10
+
+static void user_main(void* pvParameters) {
+	sled_main(0, NULL);
+}
+
+// Main entry point.
+// We can choose to pin to a core. This makes sense if your chip has more than 1 core.
+// So you can pin SLED to one core and use the other cores for stuff like WiFi, Webserver, etc...
+// Use core = -1 to not pin
+void sled_task(void* pvParameters, int core) {
+	if(xTaskGetSchedulerState() == 1)
+		vTaskStartScheduler();
+	if (core < 0 || core >= oscore_ncpus()) {
+		xTaskCreate(user_main, "sled_main", STACK_DEPTH, pvParameters, TASK_PRIORITY, NULL);
+	} else {
+		xTaskCreatePinnedToCore(user_main, "sled_main", STACK_DEPTH, pvParameters, TASK_PRIORITY, NULL, core);
+	}
+}
+
+// -- event
+oscore_event oscore_event_new(void) {
+	return NULL;
+}
+
+int oscore_event_wait_until(oscore_event ev, ulong desired_usec) {
+	portTickType waketick = oscore_udate();
+	if (waketick >= desired_usec) {
+		return waketick;
+	}
+	TickType_t wait = (desired_usec-waketick) / 1000 / 1000;
+	vTaskDelayUntil(&waketick, wait);
+	return wait;
+}
+
+void oscore_event_signal(oscore_event ev) {
+}
+
+void oscore_event_free(oscore_event ev) {
+}
+
+// Time keeping.
+// Since we don't have a RTC, presumably,
+// we'll use the uptime.
+// This aligns with the above.
+ulong oscore_udate(void) {
+	// Ticks * portTICK_RATE_MS gives us ticks in msecs, we need usecs.
+	// Even if we fake it.
+	struct timeval tv;
+	gettimeofday(&tv, NULL);
+	return (tv.tv_sec + tv.tv_usec) * 1000 * 1000;
+}
+
+
+// Below: Stubs and untestet stuff. Danger zone!
+
+// -- mutex
+oscore_mutex oscore_mutex_new(void) {
+	return xSemaphoreCreateMutex();
+}
+
+void oscore_mutex_lock(oscore_mutex m) {
+	xSemaphoreTake(m, portMAX_DELAY);
+}
+
+void oscore_mutex_unlock(oscore_mutex m) {
+	xSemaphoreGive(m);
+}
+
+void oscore_mutex_free(oscore_mutex m) {
+	free(m);
+}
+
+int oscore_ncpus(void) {
+	// TODO Use freertos API if available
+	return 2;
+}
+
+
+oscore_task oscore_task_create(const char* name, oscore_task_function func, void* ctx) {
+	TaskHandle_t xHandle = NULL;
+	xTaskCreate((void *) func, name, STACK_DEPTH, ctx, TASK_PRIORITY, &xHandle);
+	return xHandle;
+}
+
+void oscore_task_setprio(oscore_task task, int prio) {
+	if (prio < 0) {
+		prio = 0;
+	} else if (prio > configMAX_PRIORITIES -1) {
+		prio = configMAX_PRIORITIES -1;
+	}
+	vTaskPrioritySet(task, prio);
+}
+
+void oscore_task_yield(void) {
+	taskYIELD();
+};
+
+void oscore_task_pin(oscore_task task, int cpu) {}
+
+void * oscore_task_join(oscore_task task) {
+	// ye ok
```

Comment:
We probably want to implement some way of joining, but it's not required, only really for clean exits (which aren't really applicable to mcus..)

Reply:
I gave it a shot (but didn't test yet).

Reply:
It'll do for now, I think.